importccl: fix bug where GC job would not operate on any tables

pbardea opened the pull request:
Previously, IMPORT did not populate the tables to drop in the GCJob that
it created. Therefore the partially imported data of a failed or
canceled IMPORT job would not get GC'ed.

Fixes #48592.

Release note (bug fix): Previously, when an IMPORT failed, it created a
GC job which would behave as a no-op. Now the partially imported data
after an IMPORT fails or is canceled should be deleted.

pbardea committed:
importccl: fix bug where GC job would not operate on any tables

Previously, IMPORT did not populate the tables to drop in the GCJob that
it created. Therefore the partially imported data of a failed or
canceled IMPORT job would not get GC'ed.

Release note (bug fix): Previously, when an IMPORT failed, it created a
GC job which would behave as a no-op. Now the partially imported data
after an IMPORT fails or is canceled should be deleted.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48593)
<!-- Reviewable:end -->

thoszhang reviewed (commented):
Is it possible to write a test where we force a failed IMPORT and test that everything gets GC'ed?

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-:-M6pG6kA6EZP6ql-PDkV:b-lnpo8g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
oof, immediate backport. Does this mean that the table is still in namespace? I'm +1 on a test. Code looks right but this should be tested before merging. 

It seems like we may want to create a builtin to find orphaned or not dropped tables for cases like both this and https://github.com/cockroachdb/cockroach/issues/48589. As another work item, it might be a good idea to sanity check the state of the schemas and keyspace. @spaskob suggested this earlier today.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-:-M6pFyzH4El4Es-3Ugko:b-dhnsht)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

pbardea commented:
Yep, looking at ways to write a test for this now. The table descriptor will be updated to the DROP state. What this means is that it will not be present in `system.namespace`, but the descriptor in the DROP state will stay in `system.descriptor`.

pbardea committed:
importccl: add IMPORT test to exercise OnFailOrCancel

This commit adds a test that exercises IMPORT's OnFailOrCancel hook. In
particular, it ensures that a failed import will create a GC job that
will clean up any partially written data and remove it's descriptor.

Release note: None

pbardea commented:
Added a test - RFAL.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-:-M6pwNhsCvE8OzGM5NLc:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 1561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-M6pvvND9vd3jUu_kh2B:-M6pvvNE5_5w9rTPyI44:b-8j6hl1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ed8c0e632f0e66efc61a56d0240754087758b2/pkg/ccl/importccl/import_stmt_test.go#L1561)):*
> ```Go
> 			t.Fatal(err)
> 		}
> 		forceFailure = false
> ```

you already deferred setting this, don't do it again
___
*[pkg/jobs/registry.go, line 1264 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-M6pw8Fl6pVWoS87OInD:-M6pw8FmBN9ztU3vi8gz:bh5wq8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ed8c0e632f0e66efc61a56d0240754087758b2/pkg/jobs/registry.go#L1264)):*
> ```Go
> 
> // NudgeAdoptionQueueForTest is used by tests to tell the registry that there is
> // a job to be ado.
> ```

`to be ado`

@spaskob you just brought this up as an idea of something that might be handy.


<!-- Sent from Reviewable.io -->

pbardea reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48593)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/ccl/importccl/import_stmt_test.go, line 1561 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-M6pvvND9vd3jUu_kh2B:-M6q1gGe4fWI2_8xfvHb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ed8c0e632f0e66efc61a56d0240754087758b2/pkg/ccl/importccl/import_stmt_test.go#L1561)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

you already deferred setting this, don't do it again
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 1264 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-M6pw8Fl6pVWoS87OInD:-M6q1gug1911YNiZub3X:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d7ed8c0e632f0e66efc61a56d0240754087758b2/pkg/jobs/registry.go#L1264)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`to be ado`

@spaskob you just brought this up as an idea of something that might be handy.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dt commented on pkg/jobs/registry.go:
```diff
@@ -1259,3 +1259,9 @@ func (r *Registry) unregister(jobID int64) {
+
+// NudgeAdoptionQueueForTest is used by tests to tell the registry that there is
+// a job to be adopted.
+func (r *Registry) NudgeAdoptionQueueForTest() {
```
huh, github swallowed my comment, but I meant to note here that we usually _prefix_ public methods used for tests only with `Testing`, e.g. `TestingNudgeAdoptionQueue` or something

pbardea replied:
Done.

ajwerner reviewed (commented):
still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48593#-:-M78Gvkj5eimpReJ8XX9:b-5fsfav)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @dt)



<!-- Sent from Reviewable.io -->

pbardea commented:
TFTRs!
bors r=dt,ajwerner
